### ISSUE TITLE
Weird behaviour when using different IMEs with thousand seperator

### ISSUE THREAD

**Issue by HeRo3S** [s-yadav/react-number-format#843]
#### Describe the issue and the actual behavior
I tried to type 10000 to the input field using Windows' Vietnamese keyboard layout (using thousandSeperator props), but the output display incorrectly. 
![image](https://github.com/s-yadav/react-number-format/assets/71567565/6564a11a-1077-4401-8552-e1ecceec4b93)

Though the output would display correctly when I exited the input field.

#### Describe the expected behavior
The input data display correctly when user trying to input.

#### Provide a CodeSandbox link illustrating the issue
Using the documentation's sandbox here:
https://[codesandbox.io/p/sandbox/thousandsgroupstyle-demo-u3ip59?from-embed=](https://codesandbox.io/p/sandbox/thousandsgroupstyle-demo-u3ip59?from-embed=)

#### Provide steps to reproduce this issue
Use Windows' Vietnamese keyboard layout and type 10000 or more (must type 2 different numeric characters to see the bug)

#### Please check the browsers where the issue is seen
- [x] Chrome
- [ ] Chrome (Android)
- [ ] Safari (OSX)
- [ ] Safari (iOS)
- [x] Firefox
- [ ] Firefox (Android)

**Comment by tinncdev** [comment]
In your sandbox, the `thousandsGroupStyle` was set to  indian style `lakh` so the output `11,10,000` is correct
Either remove `thousandsGroupStyle` or set it to `thousand` to match your expected output.

Reference to the doc: https://s-yadav.github.io/react-number-format/docs/numeric_format/#thousandsgroupstyle-string

**Comment by RyoheiTomiyama** [comment]
I guess, 
he typed `10000`, but displayed `1110000`.

In MacOS, On MacOS, it looks like the following screenshot.
<img width="825" alt="スクリーンショット 2024-10-11 14 15 57" src="https://github.com/user-attachments/assets/a3adccd7-2f19-435f-b12a-9bcd30ccefde">

**Comment by HeRo3S** [comment]
Yeah, it only occurred so far on MS Windows with Microsoft's IME. I tried to ask a colleague who use MacOS to test it and didn't see any problem with it.
My only guess is that it's Microsoft's own way of handling input buffer in other IME that's causing the format function to kinda freak out.

**Comment by tinncdev** [comment]
Did not notice the `Windows' Vietnamese keyboard` part in previous comment.

This is related to https://github.com/s-yadav/react-number-format/issues/636

which exactly similar to [this react issue](https://github.com/facebook/react/issues/8683) which closed for duplicated to this [**9 years old issue** - Change event fires extra times before IME composition ends](https://github.com/facebook/react/issues/3926)

this was too much for my little pricing input so I ended up letting the input as is and  just use `numericFormatter` method to show a small formatted text as hint bellow the input

**Comment by sadik-malik** [comment]
Thank you, @tinncdev, for your research on this Windows IME issue affecting certain languages like Vietnamese Telex. I was able to replicate this strange input behavior across multiple libraries, including AutoNumeric.js and react-input-mask.

Even tools like VS Code seem to suffer from this issue. For example, if you type anything and then click anywhere in the editor area (outside the typed text), the text gets duplicated. Platforms like Discord also experience this problem. 

Did @microsoft ship a broken feature, or is this intended behavior?

For `react-number-format`, I managed to get things working by using a custom format function on NumberFormatBase, along with the onCompositionStart and onCompositionEnd events. In the custom format function, prevent formatting value while composition is in progress and apply formatting only after composition ends. In my opinion, this behavior is not ideal, but I haven't found a better solution yet.

```jsx
// This fix can be implemented in NumberFormatBase by library author but I don't consider this to be a perfect solution.
 const [isComposing, setIsComposing] = useState<boolean>(false);
 const _format = isComposing ? defaultFormat : (format as FormatInputValueFunction);

 const [{ formattedValue, numAsString }, onFormattedValueChange] = useInternalValues(
    propValue,
    defaultValue,
    Boolean(valueIsNumericString),
    _format,
    removeFormatting,
    onValueChange,
  );

const inputProps = Object.assign({ inputMode }, otherProps, {
    type,
    value: formattedValue,
    onChange: _onChange,
    onKeyDown: _onKeyDown,
    onMouseUp: _onMouseUp,
    onFocus: _onFocus,
    onBlur: _onBlur,
    onCompositionStart: _onCompositionStart,
    onCompositionEnd: _onCompositionEnd,
  });

// or, you can use format prop on NumberFormatBase and simply return value if composition is going on.
```

**Comment by tinncdev** [comment]
Indeed it's not ideal, composition based IME caused a lot of problem everywhere, from wrong autoformat to dupplicated words... in a lot of program
I think it better to just let the input and IME does it job and
- avoiding patching user input when typing, either do format on blur or small hint popup
-  for some special cases, use a virtual input (popover dial pad like some banking app does) to avoid native keyboard input events
This still be a good note for those who really need this behaviour and have a big composition based IME user base
